coding/coding.go

```diff
@@ -1,75 +1,112 @@
-package ipfsld
+package coding
 
 import (
-	mc "github.com/jbenet/go-multicodec"
-	mccbor "github.com/jbenet/go-multicodec/cbor"
-	mcmux "github.com/jbenet/go-multicodec/mux"
+	"bytes"
+	"fmt"
+	"io"
 
-	ipld "github.com/ipfs/go-ipld"
+	cbor "github.com/ipfs/go-ipld/coding/cbor"
+	json "github.com/ipfs/go-ipld/coding/json"
 	pb "github.com/ipfs/go-ipld/coding/pb"
+
+	memory "github.com/ipfs/go-ipld/memory"
+	stream "github.com/ipfs/go-ipld/stream"
+	mc "github.com/jbenet/go-multicodec"
+)
+
+var Header []byte
+
+const (
+	HeaderPath = "/mdagv1"
 )
 
-// defaultCodec is the default applied if user does not specify a codec.
-// Most new objects will never specify a codec. We track the codecs with
-// the object so that multiple people using the same object will continue
-// to marshal using the same codec. the only reason this is important is
-// that the hashes must be the same.
-var defaultCodec string
+type Codec int
 
-var muxCodec *mcmux.Multicodec
+const (
+	NoCodec       Codec = 0
+	CodecProtobuf Codec = iota
+	CodecCBOR
+	CodecJSON
+	CodecCBORNoTags
+)
+
+var StreamCodecs map[string]func(io.ReadSeeker) (stream.NodeReader, error)
```

Comment:
btw, this whole module should be using `multicodec` and the coding pattern establish there. i created the `Codec` abstractions for this use case. they achieve something similar.